### ISSUE TITLE
关于DSAC或GOPS的状态空间的特征提取网络的扩展问题。

### ISSUE THREAD

**Issue by xiaolong0912** [Jingliang-Duan/DSAC-v2#14]
现在的特征提取网络似乎只是单一的MLP/CNN/CNN_SHARED/RNN/POLY/GAUSS这几种类型，但是在较为复杂的控制场景，例如自动驾驶车辆端到端控制中，输入特征往往即包含图像网络，也包含主车速度和加速度，以及规划路径点等离散值，是多种类型的混合输入。
根据gym官方的写法，我们可以将每种输入特征使用 gym.spaces.Box进行包装，随后返回状态空间为gym.spaces.Dict类型，根据字典键值类型进行自定义特征提取，随后进行拼接操作。
请问能否扩展一下状态空间的输入源类型，让其能够支持gym.spaces.Dict类型。或者我们能否对模型的决策网络或者AC特征提取网络进行自定义修改？类似stablebaseline3中的Multiple Inputs and Dictionary Observations，自定义特征提取器。

万分感谢！！！

下面给出SB3库中的示例代码，类似这样的实现方式：
import gymnasium as gym
import torch as th
from torch import nn

from stable_baselines3.common.torch_layers import BaseFeaturesExtractor

class CustomCombinedExtractor(BaseFeaturesExtractor):
    def __init__(self, observation_space: gym.spaces.Dict):
        # We do not know features-dim here before going over all the items,
        # so put something dummy for now. PyTorch requires calling
        # nn.Module.__init__ before adding modules
        super().__init__(observation_space, features_dim=1)

        extractors = {}

        total_concat_size = 0
        # We need to know size of the output of this extractor,
        # so go over all the spaces and compute output feature sizes
        for key, subspace in observation_space.spaces.items():
            if key == "image":
                # We will just downsample one channel of the image by 4x4 and flatten.
                # Assume the image is single-channel (subspace.shape[0] == 0)
                extractors[key] = nn.Sequential(nn.MaxPool2d(4), nn.Flatten())
                total_concat_size += subspace.shape[1] // 4 * subspace.shape[2] // 4
            elif key == "vector":
                # Run through a simple MLP
                extractors[key] = nn.Linear(subspace.shape[0], 16)
                total_concat_size += 16

        self.extractors = nn.ModuleDict(extractors)

        # Update the features dim manually
        self._features_dim = total_concat_size

    def forward(self, observations) -> th.Tensor:
        encoded_tensor_list = []

        # self.extractors contain nn.Modules that do all the processing.
        for key, extractor in self.extractors.items():
            encoded_tensor_list.append(extractor(observations[key]))
        # Return a (B, self._features_dim) PyTorch tensor, where B is batch dimension.
        return th.cat(encoded_tensor_list, dim=1)

**Comment by xiaolong0912** [comment]
我所使用的强化学习环境为自定义的carla环境，其中step()方法返回的状态空间为gym.spaces.Dict类型

**Comment by xiaolong0912** [comment]
我已经实现了自适应的特征提取网络，只需要保证给GOPS的状态空间返回是一个完整的numpy变量即可。在给GOPS返回状态空间之前，使用自定义的特征提取网络提前对状态空间进行特征提取和拼接融合处理，后续的AC网络仍旧使用MLP类型。

**Comment by xiaolong0912** [comment]
请问，除了我目前解决的自定义特征提取器，GOPS是否支持自定义特征网络，是否支持多个输入和字典类型的状态空间？感谢！

**Comment by drsssssss** [comment]
> 请问，除了我目前解决的自定义特征提取器，GOPS是否支持自定义特征网络，是否支持多个输入和字典类型的状态空间？感谢！

dict的obs space做过，周末可以整理下推个仓库供参考，多个输入是什么意思？自定义特征网络是指obs要过一遍encoder吗

**Comment by drsssssss** [comment]
关于”可满足多输入的字典类型的obs需求“问题回复：

首先，感谢您的建议！GOPS开发团队认为这是一个很好的通用性建议，相关功能也正在开发中；其次，现有GOPS是支持自定义网络架构的，仅需继承现有的网络基类开发即可；”更多的线性层“问题仅需调整example_train中训练脚本中hidden_sizes相关参数即可实现。

最后，再次表示感谢，欢迎持续关注GOPS官方仓库更新状况！
